docs/proposals/20210712-sidecarsetimagespullsecrets.md

```diff
@@ -0,0 +1,138 @@
+---
+title: SidecarSetImagePullSecrets
+
+authors:
+- "@veophi"
+  
+reviewers:
+- "@Fei-Guo"
+- "@furykerry"
+- "@FillZpp"
+  
+creation-date: 2021-07-12
+  
+last-updated: 2021-07-12
+  
+status: implementable
+
+---
+# SidecarSetImagePullSecrets
+
+- Provide a way to pull images using `Secret` from private repositories for sidecarSet.
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+* [SidecarSetImagePullSecrets](#sidecarsetimagepullsecrets)
+    * [Motivation](#motivation)
+    * [Proposal](#proposal)
+        * [Scheme #1: Create Secret By SidecarSet](#scheme-1-create-secret-by-sidecarset)
+            * [Advantages](#advantages)
+            * [Problems](#problems)
+        * [Scheme #2: Create Secret By User](#scheme-2-create-secret-by-user)
+            * [Advantages](#advantages-1)
+            * [Problems](#problems-1)
+    * [Implementation History](#implementation-history)
+
+Created by [gh-md-toc](https://github.com/ekalinin/github-markdown-toc)
+
+## Motivation
+
+In k8s, secret is an easy and safe way to store and manage sensitive information, such as passwords, OAuth tokens, and ssh keys.
+One of the most common-used features of it is to pull images from private repositories. However, sidecarSet does not support this feature so far.
+
+## Proposal
+
+When a sidecarSet is injected into the matched `Namespace.Pod`, the references of the secrets it needs also will be injected into the pod. 
+In this situation, we have to consider the following two questions:
+1. Which `Secret` we should use? I mean for the required `Namespace.Secret`, how to decide the `Namespace` when using **cluster-level sidecarSet**?
+2. Does the `Namespace` have the `Secret`?
+
+Considering the two questions above, we provide two schemes to discuss:
+(I prefer the second.)
+
+### Scheme #1: Create Secret By SidecarSet
+**Main idea:**
+  - Which `Secret` we should use?  **We Create the Secret for ALL namespaces!**
+  - Does the `Namespace` have the `Secret`? **We Create the Secret for ALL namespaces!**
+
+This is a rough but effective way to solve these two problems. 
+
+First, we add the secret fields in `sidecarset_types.go`, it looks like this:
+```go
+type SidecarSetSpec struct {
+	...
+	SecretFields []SecretField
+}
+
+type SecretField struct {
```

Comment:
If you want to put the secrets definition in SidecarSet, the data and name are not enough. Actually the `data` is a map and there should also have a `type` field. You may need to define `SecretTemplates []v1.Secret` in `SidecarSetSpec`.

Another question is that users should define the real data or the data after base64 encoding here?